lib/puppet/provider/user/useradd.rb

```diff
@@ -147,19 +147,47 @@ def add_properties
     # validproperties is a list of properties in undefined order
     # sort them to have a predictable command line in tests
     Puppet::Type.type(:user).validproperties.sort.each do |property|
-      next if property == :ensure
-      next if property_manages_password_age?(property)
-      next if property == :groups and @resource.forcelocal?
-      next if property == :expiry and @resource.forcelocal?
-      # the value needs to be quoted, mostly because -c might
-      # have spaces in it
-      if value = @resource.should(property) and value != ""
-        cmd << flag(property) << munge(property, value)
-      end
+      value = get_value_for_property(property)
+      next if value.nil?
+      cmd << flag(property) << munge(property, value)
     end
     cmd
   end
 
+  def get_value_for_property(property)
+    return nil if property == :ensure
+    return nil if property_manages_password_age?(property)
+    return nil if property == :groups and @resource.forcelocal?
+    return nil if property == :expiry and @resource.forcelocal?
+    # the value needs to be quoted, mostly because -c might
+    # have spaces in it
+    value = @resource.should(property)
+    return nil if !value || value == ""
+
+    value
+  end
+
+  def has_sensitive_data?(property = nil)
+    properties = property ? [property] : Puppet::Type.type(:user).validproperties
```

Comment:
Does the `user` type contain any sensitive parameters (not properties)? If so, we may need to call `validparameters` instead.

Reply:
Are properties a subset of parameters?

Reply:
Yeah, [`Puppet::Property` subclasses `Puppet::Parameter`](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/property.rb#L41). The former are managed attributes (like `password`). The latter are not managed, but contribute additional information about how to manage the resource, like a file's `source`. I don't remember if sensitive only works with properties, or all parameters?

Reply:
Hmm, this seems to only return parameter names, not objects:
```
properties = property ? [property] : Puppet::Type.type(:user).parameters
```

Reply:
I think calling `validproperties` is ok for the user type, but I wanted to double check. Are any of the parameters sensitive? I assume not?

```
> Puppet::Type.type(:user).parameters
=> [:name, :membership, :system, :allowdupe, :managehome, :role_membership, :auth_membership, :profile_membership, :key_membership, :ia_load_module, :attribute_membership, :forcelocal, :purge_ssh_keys, :provider]
```

Reply:
We can discuss this in more detail, but I wasn't sure how to get the parameter objects (not just the names) and I wasn't sure if it was worth adding access code just for this PR.

Reply:
You can get the parameter (or its derived property) using:

```ruby
 user.parameter(:system)
```

but it only works if that param has been specified. That said it doesn't appear any parameters can be sensitive, so marking this as resolved.